sources/driverdog/src/main.rs

```diff
@@ -0,0 +1,359 @@
+                name,
+                shared_object,
+                &build_dir,
+                driver_path,
+            )?);
+        }
+    }
+
+    // Generete paths to kernel module's dependencies
+    let mut dependencies = kernel_module
+        .dependencies
+        .iter()
+        .map(|d| driver_path.join(d).to_string_lossy().into_owned())
+        .collect();
+
+    // Link the kernel module
+    let mut args = vec![
+        "-r".to_string(),
+        "-T".to_string(),
+        common_module_script_path.display().to_string(),
+        "--build-id".to_string(),
+        "-o".to_string(),
+        kernel_module_path,
+    ];
+
+    args.append(&mut runtime_dependencies_paths);
+    args.append(&mut dependencies);
+    command(&LD_BIN_PATH, &args)?;
+    info!("Linked {}.ko", name);
+
+    Ok(())
+}
+
+/// Links the given shared object
+fn link_shared_object<P, B, S>(
+    name: S,
+    shared_object: &SharedObject,
+    build_dir: B,
+    driver_path: P,
+) -> Result<String>
+where
+    S: AsRef<str>,
+    B: AsRef<Path>,
+    P: AsRef<Path>,
+{
+    let name = name.as_ref();
+    let build_dir = build_dir.as_ref();
+    let driver_path = driver_path.as_ref();
+
+    // Temporary files are created in build_dir
+    let object_path = Path::new(build_dir)
+        .join(format!("{}.o", name))
+        .to_string_lossy()
+        .into_owned();
+    // Paths to the dependencies for this shared object
+    let mut dependencies = shared_object
+        .dependencies
+        .iter()
+        .map(|d| {
+            Path::new(driver_path)
+                .join(d)
+                .to_string_lossy()
+                .into_owned()
+        })
+        .collect();
+
+    // Link the shared object
+    let mut args = vec!["-r".to_string(), "-o".to_string(), object_path.clone()];
+    args.append(&mut dependencies);
+
+    command(&LD_BIN_PATH, &args)?;
+    info!("Linked object '{}.o'", name);
+
+    // Strip the shared object
+    command(
+        &STRIP_BIN_PATH,
+        &[
+            "-g",
+            "--strip-unneeded",
+            "--keep-symbol",
+            "init_module",
+            "--keep-symbol",
+            "cleanup_module",
+            &object_path,
+        ],
+    )?;
+    info!("Stripped object '{}'.o", name);
+
+    Ok(object_path)
+}
+
+/// Loads the linked kernel modules
+fn load_kernel_modules(driver_config: &DriverConfig) -> Result<()> {
+    let mut kernel_modules: Vec<String> = driver_config
+        .kernel_modules
+        .keys()
+        .map(|k| k.to_string())
+        .collect();
+
+    // Update the modules.dep with the newly created kernel modules
+    let args: Vec<String> = Vec::new();
+    command(DEPMOD_BIN_PATH, &args)?;
+    info!("Updated modules dependencies");
+
+    // Load kernel modules
+    let mut args = vec!["-a".to_string()];
+    args.append(&mut kernel_modules);
+    command(MODPROBE_BIN_PATH, &args)?;
+    info!("Loaded kernel modules");
```

Comment:
Linking the modules makes sense to me, but I'm not sure about running `modprobe`, and even `depmod` is a gray area. My misgiving is that we're basically managing global state but doing it from a place where we can't know it's safe - there might be another `driverdog` process running, or we might've loaded the modules previously and that will fail, or `depmod` might encounter some transient files from a kmod build in progress and bomb out.

When we look at adding support for multiple versions of drivers, a workflow like this might make sense:
* spawn `driverdog` and have it link all modules at once, writing to versioned output directories
* assuming the modules can't all be installed under `/lib/modules` because of filename conflicts, we could add a symlink from `/lib/modules/.../kernel/drivers/extra/video/nvidia -> /var/lib/nvidia/driver`, and another symlink from `/var/lib/nvidia/driver -> tesla`.
* every time we update that second symlink, we'd re-run `depmod`
* unloading and reloading a set of nvidia drivers seems fraught with complications, but for the initial install then it could be just `modprobe nvidia`. Or we might regenerate a `modprobe.d` config file and then call `modprobe nvidia`.

Whether or not that's the direction we end up taking, I feel like putting the functionality here is inviting us to expand the charter of this helper program. The alternative is that we run `driverdog`, `depmod`, and `modprobe` in quick succession which seems equally easy to do today, and doesn't lock us into an architectural direction.